### DISCUSSION TITLE
How to test event choreography?

### DISCUSSION THREAD

**Discussion by danstooamerican** [spring-projects/spring-modulith#96] [General]
Hi,

I was wondering how I can write integration tests that verify a business process that is driven by application events (a saga essentially).
For example, the creation of a new account triggers an invitation. After the invitation is accepted the account is activated.
The corresponding events are `AccountCreationInitiated`, `InvitationAccepted`, and `AccountActivated`.

I want to be able to write a test, that:

1. Creates a new account
2. Gets hold of the invitation (ideally through the published event because the invitation process is handled by another module)
3. Accepts the invitation
4. Verifies that `AccountActivated` was published

I am able to write steps 1, 3, and 4 but step 2 is tricky. The problem is, that step two is performed by an `@ApplicationModuleListener` . This event handler is never called inside my test (it works if I start my application normally). I think it is because it waits for the transaction to finish but this never happens inside my test I think. The test runs fine if I replace the `@ApplicationModuleListener` with `@EventHandler`.

I am also not sure how the `@Async` annotation inside `@ApplicationModuleListener` interferes with my tests and how to reliably test these event handlers. I already added delays and timeouts to my verify calls which did not change the behavior.

Lastly, I would like to know if it makes sense to publish events inside my test methods to see how my system behaves.

How should I go about writing my integration tests?

**Comment by odrotbohm** [comment]
That's an interesting question. The natural scope in focus for Spring Modulith is indeed a module in isolation. That's what our [integration test support](https://docs.spring.io/spring-modulith/docs/current/reference/html/#testing) is centered around. That said, there's also the `PublishedEvents` abstraction (docs still missing 😬) that can be used to capture application events published during the execution of a test case. With e.g. [Awaitility](https://www.baeldung.com/awaitility-testing), that could wait until the `PublishedEvents` instance has seen the `AccountActivated` event.

I wouldn't mind hearing about the results you can achieve with that. `PublishedEvents`' API is currently pretty basic. I can imagine we might want to add a few more dedicated methods like `….hasSeenEventOfType(Class<?>)` so that you could end the Awaitility pipeline with `….until(() -> events.hasSeenEventOfType(…))`.

**Reply by odrotbohm** [reply]
Just to give you a heads up: the 0.3 snapshots now contain an `AssertablePublishedEvents` for easier usage with AssertJ. See the [reference documentation](https://docs.spring.io/spring-modulith/docs/current-SNAPSHOT/reference/html/#events.integration-testing) for details. Also, `PublishedEvents` now exposes an `….eventOfTypeWasPublished(…)` for convenient use with Awaitility.

**Reply by danstooamerican** [reply]
Thank you, I will look into it.

Maybe as an inspiration for the PublishedEvents interface: Axon Framework has very nice test fixtures which offer a `given, when, then` like interface. Basically, the fixture lets users publish events and commands, and easily assert what events were observed.

**Reply by danstooamerican** [reply]
Ok, so currently I don't see how I can test the following scenario in my integration tests. Maybe @odrotbohm can shed some light on it.

I have two modules that are supposed to interact through events. The process starts in `Module A` which publishes an event:

```
public void handle(SomeCommand command) {
    // do meaningful work...

    events.publishEvent(new SomethingImportantHappened(...));
}
```

The `SomethingImportantHappened` event is picked up by `Module B`:

```
@ApplicationModuleListener
public void on(SomethingImportantHappened event) {
    // do meaningful work...

    events.publishEvent(new SomethingImportantWasHandled(...));
}
```

After `SomethingImportantWasHandled` is published, `Module A` finalizes the process:

```
@ApplicationModuleListener
public void on(SomethingImportantWasHandled event) {
    // finalize process...
}
```

This works without issues when I run my application but is impossible to test at the moment. My test is supposed to test `Modul B` in isolation. My scenario is: "After `SomethingImportantHappened` is published, I expect `SomethingImportantWasHandled` to be published as a response".

My test:

```
@Test
void my_integration_test(PublishedEvents publishedEvents) {
    events.publishEvent(new SomethingImportantHappened(...));

    // could be written cleaner with your 0.3 adaptations. Tbh I still wish it would be simpler like expectEvents(new SomethingImportantWasHandled(...))
    await().until(() -> publishedEvents.ofType(SomethingImportantWasHandled.class).iterator().hasNext());
}
```

I believe the problem is related to the `@ApplicationModuleListener` not being triggered without a transaction. I tried to work around it but with no success:

```
@Test
@Transactional // this is necessary because TestTransaction.start() does not find a TransactionContext without it
void my_integration_test_with_transaction(PublishedEvents publishedEvents) {
    events.publishEvent(new SomethingImportantHappened(...));

    // try to invoke @ApplicationModuleListener by committing the transaction
    TestTransaction.end();

    await().until(() -> publishedEvents.ofType(SomethingImportantWasHandled.class).iterator().hasNext());
}
```

The test succeeds if I replace `@ApplicationModuleListener` with `@EventListener` but this is not desired.

**Reply by odrotbohm** [reply]
Does it make sense to avoid using the publication infrastructure to implement "if SomethingImportantHappened is published"? Doesn't "SomethingImportantHappened is published" equal to the method invocation of the listener for it? In other words, why not simply call the event listener directly as then triggers the downstream functionality.

With your approach, you're not only including the publication mechanism in the test (which is a bit of expansion of the test scope), you then of course also have to adhere to the mechanism of the publication, which means: running the publication from inside a transaction. The easiest way to achieve that is by using `TransactionOperations`:

```java
@ApplicationModuleTest
@RequiredArgsConstructor
public class SomeIntegrationTests {

  private final ApplicationEventPublisher events;
  private final TransactionOperations tx;

  @Test
  void testName() {

    tx.executeWithoutResult(it -> {
      events.publishEvent(…);
    });

    …
  }
}
```

Does that help?

Oh, and a minor comment: IIRC you can simply return event objects from event listener methods, and they will be published to the event bus again. So, no need to depend on `ApplicationEventPublisher` and actively publish an event yourself.

**Reply by danstooamerican** [reply]
Ok, this is a fair point. I have adapted my test to this:

```
@Test
void testName(PublishedEvents publishedEvents) {
    myservice.on(new SomethingImportantHappened(...));

    // tried using Awaitility, but the event is never observed
    await().until(() -> publishedEvents.ofType(SomethingImportantWasHandled.class).iterator().hasNext());

    // normal method does not work either
    assertThat(publishedEvents.ofType(SomethingImportantWasHandled.class))
            .areExactly(1, new Condition<>(evt -> ...);
}
```

Sadly, the `SomethingImportantWasHandled` event is never observed. The `myservice.on(...)` method is annotated with `@ApplicationModuleListener`. If I remove the annotation or change it to `@EventListener`, the test succeeds.

**Reply by odrotbohm** [reply]
I think we have discovered a bug. 😿  `PublishedEvents` is tied to a thread-bound `ApplicationListener` that will only see events from the thread that invokes the test method. Given that we're incentivizing users to rather resort to asynchronous integration of modules, that's not terribly useful. Also, it turns out that, contrary to what I suggested above, `@Async` methods cannot return event objects directly but would have to wrap those into a `CompletableFuture`.

I'll take this to the team and especially discuss the thread bound nature of `PublishedEvents`. Switching to a singleton instance might severely affect the ability to run tests in parallel, but we'll have to discuss how much of an issue that is.

Thanks for your patience and the very useful feedback so far!

**Reply by danstooamerican** [reply]
Always happy to help. We are looking forward to a fix :)

**Reply by odrotbohm** [reply]
I've filed and fixed GH-116. `PublishedEvents` should now see events published from all threads directly or indirectly spawned from the main application thread. Please let me know how it goes.

**Reply by danstooamerican** [reply]
Wow, that was super fast. Thank you!

My test case succeeds now. I'll report back if I hit any bumps.

Btw I can now also use the event publication infrastructure in my test without problems. Somehow it feels more natural to me to write the integration test like this since the public API of the module is to act on the event publication. The specific handler is an implementation detail from my point of view. What are your thoughts?

**Reply by odrotbohm** [reply]
That is a good point. I think it would be just fine to start with the transactional event publication, as that indeed requires less knowledge about how to start the flow. You just kick it off, defining which state your module is in and which event actually stimulates it. That's very much in line with the encapsulating nature of a module.

I was wondering if, at the end, we might come up with some API that makes formulating these tests a bit less cluttered. Currently, you have to interact with `TransactionOperations`, `ApplicationEventPublisher` and `Awaitility` all separate from each other, get the former two injected into your test case. Maybe we can use the `AssertablePublishedEvents` to expose a factory method that would allow a fluent setup of transactional event publication, await condition plus a final code block to verify assertions once the condition has been met. Similar to the way assertions work in AssertJ, but slightly more high-level for the setup phase.

To drive that, experiments like yours are really valuable!

**Reply by danstooamerican** [reply]
That is exactly what I was thinking about. I worked with Axon Framework in the past which offers very nice testing support.
You might want to have a look at their [documentation ](https://docs.axoniq.io/reference-guide/axon-framework/testing/commands-events) to get some inspiration. I especially like the `given, when, then` API and the helper functions for assertions (this could use AssertJ internally). The assertions of the `PublishedEvents` feel quite cumbersome at the moment because I always have to define the `Conditions`. I find their clear separation into Commands, Queries, and Events very helpful. We use Modulith with this idea in the back of our minds and it works great so far.

They call their test helpers `TestFixtures` which I think aptly describes the concept. I would personally prefer something along this terminology over `PublishedEvents`.

**Reply by odrotbohm** [reply]
I don't think we're going to go anywhere near what Axon provides. We're not in the business of CQRS. All we care about is events, as they allow us to decouple the modules. Whether you introduce commands to trigger functionality or just invoke methods on beans listening to these events is not something we care about. I guess if you want to buy into CQRS (which is perfectly fine) than you'd just go with a framework that supports that. There's no real point in us trying to replicate what's already existing in a well-supported and -designed way in Axon.  Also, from what I hear, the additional technical complexity of those frameworks is also something that puts people off. In the end, I guess it's a tradeoff.

I guess we could add a couple of convenience methods to `AssertablePublishedEvents` to, for example, expect event *instances* matching, setting up waiting conditions etc.

**Reply by danstooamerican** [reply]
I did not mean to imply that Modulith should replicate Axon or any CQRS features. Those patterns have its place and purpose. I specifically did not choose Axon for our project because I needed something more lightweight but also supported an event driven programming style. So far Modulith has been minimally invasive and very easy to work with. We enjoy working with Modulith. What I meant to say is that the Axon `TestFixtures` are something I really enjoyed working with so maybe some ideas can be borrowed. 

```
@Test
public void testName() {
    fixture.given(() -> {
        // put some logic to be executed during the Arrange step of the test (i.e., call module interface)
    })
    .when(() -> {
        // Act step of the test
     })
     // .whenEventIsPublished(new SomethingImportantHappened(...)) as a helper to publish events
     .assertThat(...) // some interface to put asserts
     .expectEvents(new SomethingImportantHappened(....)); // interface to assert event publication
}
```

That is what I have in mind for an improved `PublishedEvents` interface.

**Reply by odrotbohm** [reply]
I just filed #136 and pushed a first draft of an API. A test method parameter type of `Scenario` is now available to build integration testing descriptions. Would love to hear what you think.
